src/fandango/evolution/adaptation.py

```diff
@@ -36,71 +41,90 @@ def update_parameters(
         evaluator: Evaluator,
         current_max_repetition: int,
     ) -> tuple[float, float]:
+
         diversities = evaluator.compute_diversity_bonus(population)
         avg_diversity = sum(diversities) / len(diversities) if diversities else 0
 
-        fitness_improvement_threshold = (
-            0.01  # minimal improvement to be considered significant
-        )
-        diversity_low_threshold = 0.1  # low diversity threshold
+        if prev_best_fitness > 0:
+            fitness_improvement = (
+                current_best_fitness - prev_best_fitness
+            ) / prev_best_fitness
+        else:
+            fitness_improvement = current_best_fitness
 
-        # Adaptive Mutation
+        fitness_improvement_threshold = 0.005
+        diversity_low_threshold = 0.1
+
+        # Mutation rate adjustment
         if (
-            current_best_fitness - prev_best_fitness
-        ) < fitness_improvement_threshold or avg_diversity < diversity_low_threshold:
+            fitness_improvement < fitness_improvement_threshold
+            or avg_diversity < diversity_low_threshold
+        ):
             new_mutation_rate = min(1.0, self.mutation_rate * 1.1)
             LOGGER.info(
-                f"Generation {generation}: Increasing mutation rate from {self.mutation_rate:.2f} to {new_mutation_rate:.2f}"
+                f"Generation {generation}: Increasing mutation rate from {self.mutation_rate:.3f} to {new_mutation_rate:.3f}"
             )
             self.mutation_rate = new_mutation_rate
         else:
             new_mutation_rate = max(0.01, self.mutation_rate * 0.95)
             LOGGER.info(
-                f"Generation {generation}: Decreasing mutation rate from {self.mutation_rate:.2f} to {new_mutation_rate:.2f}"
+                f"Generation {generation}: Decreasing mutation rate from {self.mutation_rate:.3f} to {new_mutation_rate:.3f}"
             )
             self.mutation_rate = new_mutation_rate
 
-        # Adaptive Crossover
+        # Crossover rate adjustment
         if avg_diversity < diversity_low_threshold:
-            new_crossover_rate = min(0.9, self.crossover_rate * 1.05)
+            new_crossover_rate = max(0.1, self.crossover_rate * 0.95)
```

Comment:
Nice catch! Makes sense in my mind: If the population is not diverse enough, decrease crossover rate to prevent inbreeding.

But also: Did you talk to whoever implemented this first to check if there's a reason they did it the other way around?